Upstream patch dwm-togglefloatingcenter-20210806-138b405f.diff (https://dwm.suckless.org/patches/togglefloatingcenter/),
applied in this order to dwm 6.5.

diff --git a/dwm.c b/dwm.c
index 65d0939..94868f9 100644
--- a/dwm.c
+++ b/dwm.c
@@ -1872,6 +1872,10 @@ togglefloating(const Arg *arg)
 	if (selmon->sel->isfloating)
 		resize(selmon->sel, selmon->sel->x, selmon->sel->y,
 			selmon->sel->w, selmon->sel->h, 0);
+
+    selmon->sel->x = selmon->sel->mon->mx + (selmon->sel->mon->mw - WIDTH(selmon->sel)) / 2;
+    selmon->sel->y = selmon->sel->mon->my + (selmon->sel->mon->mh - HEIGHT(selmon->sel)) / 2;
+
 	arrange(selmon);
 }
 
